Initial commit of shanghai hardfork updates.

philip-morlier opened the pull request.

philip-morlier committed:
Initial commit of shanghai hardfork updates.

AusIV commented on api/utils.go:
```diff
@@ -251,6 +254,21 @@ func getBlocks(ctx context.Context, db *sql.DB, includeTxs bool, chainid uint64,
+		var withdrawals []map[string]interface{}
+		switch {
+		case len(withdrawalHashBytes) == 0:
+			continue
```
You don't want to continue here, that will result in this block not getting appended to the list. You can have a case with no content, which I think is what we want here (or you could just have a comment explaining what this case represents - blocks prior to the introduction of withdrawals.

AusIV commented on api/utils.go:
```diff
@@ -846,3 +867,49 @@ func getFlumeTransactionReceiptsBlock(ctx context.Context, db *sql.DB, offset, l
+// PRIMARY KEY (block, wtdrlIndex)
+
+func getWithdrawals(ctx context.Context, db *sql.DB, block uint64) ([]map[string]interface{}, error) {
+	query := fmt.Sprintf("SELECT withdrawls.wtdrlIndex, withdrawls.vldtrIndex, withdrawls.recipient, withdrawls.amount WHERE withdrawls.block = %v;", block)
```
You need a `FROM` clause here, ie. `SELECT fields FROM withdrawals WHERE withdrawals.block = ...`

AusIV replied:
But also, it would be better to do `WHERE withdrawals.block = ?` and then add the block number as a parameter when you run `QueryContext`.

In general, formulating SQL statements with `Sprintf` needs to be done very carefully, and never with data provided to you by a user. https://xkcd.com/327/

AusIV replied:
Lastly, "withdrawals" is misspelled multiple times in this line, while spelled correctly in the migrations file.

AusIV replied:
Let's try to get a couple of simple tests going here that could at least exercise the SQL statements. I can help you craft some sample data, since we don't yet have a real chain to pull from.

philip-morlier committed:
Migration, indexing, and block API all modified for shanghai.

There is still some work to be done with the tests in indexer/ and polygon/.
Also the modified blocks.sqlite in testing-resources has not been commit as it is unclear if the testing problems originate there.

philip-morlier committed:
Polygon tests fixed, indexer stable.

philip-morlier committed:
As is commit, indexer test broken, need to rework connectToTestDBs().

philip-morlier committed:
Functional feature, all tests passing.

philip-morlier committed:
Removed debugging comments and added explanatory comments.

AusIV commented on indexer/block_indexer.go:
```diff
@@ -98,7 +109,23 @@ func (indexer *BlockIndexer) Index(pb *delivery.PendingBatch) ([]string, error)
 	uncleRLP, _ := rlp.EncodeToBytes(uncles)
 	statements := []string{
-		ApplyParameters("DELETE FROM blocks WHERE number >= %v", pb.Number),
+		ApplyParameters("DELETE FROM blocks WHERE number >= %v", pb.Number), ApplyParameters("DELETE FROM withdrawals WHERE block >= %v", pb.Number),
```
Just from a cleanliness perspective, let's break this into two lines where each entry in the list becomes one line.

AusIV commented:
Jut the one small note. Aside from that it looks good. I want to wait to merge this until a Geth release is made that will support Shanghai on the first testnets, in case anything significant changes.

philip-morlier committed:
split delete statments into two lines, block indexer.

philip-morlier committed:
updated evm to 1.4.0

philip-morlier committed:
fixed typo in compat plugin

philip-morlier committed:
merged develop into fetature branch to prepare for pull request.
